# HG changeset 9673 patch
From: kaf24@firebug.cl.cam.ac.uk
# Node ID 09dbe36de14491f51f33b33ff7c7c8ea04937355
# Parent  5da82db00c6e6dfa8dd33809c8fab94376635ad3
Subject: fix netfront/netback skb field initialization
When we copy packets in netback/netfront make sure the new skb has
all the necessary fields initialised. In particular, before we were
not copying ip_summed and that screws up checksum offload.

Signed-off-by: Keir Fraser <keir@xensource.com>
xen-unstable changeset:   9972:91c77df11b43894a2940b03029be46f6a1c85319
xen-unstable date:        Wed May 10 17:30:42 2006 +0100

Minus the stuff 9671 (unstable 9969) did and this one reverted.

Acked-By: Jan Beulich <jbeulich@novell.com>
References: 17xxxx

Index: head-2006-05-11/drivers/xen/netback/netback.c
===================================================================
--- head-2006-05-11.orig/drivers/xen/netback/netback.c	2006-05-12 09:25:37.000000000 +0200
+++ head-2006-05-11/drivers/xen/netback/netback.c	2006-05-12 09:26:11.000000000 +0200
@@ -170,7 +170,9 @@ int netif_be_start_xmit(struct sk_buff *
 		ret = skb_copy_bits(skb, -hlen, nskb->data - hlen,
 				     skb->len + hlen);
 		BUG_ON(ret);
+		/* Copy only the header fields we use in this driver. */
 		nskb->dev = skb->dev;
+		nskb->ip_summed = skb->ip_summed;
 		nskb->proto_data_valid = skb->proto_data_valid;
 		dev_kfree_skb(skb);
 		skb = nskb;
Index: head-2006-05-11/drivers/xen/netfront/netfront.c
===================================================================
--- head-2006-05-11.orig/drivers/xen/netfront/netfront.c	2006-05-12 09:25:37.000000000 +0200
+++ head-2006-05-11/drivers/xen/netfront/netfront.c	2006-05-12 09:26:11.000000000 +0200
@@ -663,7 +663,10 @@ static int network_start_xmit(struct sk_
 			goto drop;
 		skb_put(nskb, skb->len);
 		memcpy(nskb->data, skb->data, skb->len);
+		/* Copy only the header fields we use in this driver. */
 		nskb->dev = skb->dev;
+		nskb->ip_summed = skb->ip_summed;
+		nskb->proto_data_valid = skb->proto_data_valid;
 		dev_kfree_skb(skb);
 		skb = nskb;
 	}
@@ -895,8 +898,11 @@ static int netif_poll(struct net_device 
 				skb_reserve(nskb, 2);
 				skb_put(nskb, skb->len);
 				memcpy(nskb->data, skb->data, skb->len);
+				/* Copy any other fields we already set up. */
 				nskb->dev = skb->dev;
 				nskb->ip_summed = skb->ip_summed;
+				nskb->proto_data_valid = skb->proto_data_valid;
+				nskb->proto_csum_blank = skb->proto_csum_blank;
 			}
 
 			/* Reinitialise and then destroy the old skbuff. */
